aip/0194.md

```diff
@@ -0,0 +1,237 @@
+---
+aip:
+  id: 194
+  state: reviewing
+  created: 2019-08-16
+permalink: /194
+redirect_from:
+  - /0194
+---
+
+# Retrying for client libraries
+
+This document aims to lay out guidance about which requests to retry, and for
+which codes.
+
+Note: this document often uses the term "CRUD". This acronym stands for
+Create, Read (Get, List, etc.), Update, Delete.
+
+## Guidance
+
+Requests can be categorized as either unary or streaming. A unary request is
+a single request with a single response. A streaming request involves multiple
+pieces of data being sent.
+
+Unary requests can be further categorized as:
+
+- Retryable: requests in which repeated runs would not cause unintended state
+  changes. For example: most CRUD RPCs which take a unique identifier are
+  considered retryable (see rationale below in "Idempotency and pragmaticism").
+- Transactional: requests which are part of a transaction. These requests
+  **should not** be retried individually, but instead **should** have user-level
+  retry logic that retries the entire transaction block from the start.
+- Non-retryable: requests which **should not** be retried. Examples include
+  CRUD operations which don't take a unique ID, most SQL operations, and
+  operations that alter state when retried (for example,
+  `IncrementBy(some-uuid, some-field, 7)`, which takes a unique ID but will
+  always alter state).
+
+Streaming requests, on the other hand, can be further categorized as:
+
+- Server streaming: requests in which the server streams data to the client.
+  These involve no data sent from the client, meaning that it generally is an
+  idempotent `GET`. That makes it equivalent to a retryable unary request as far
+  as retrying error codes go.
+- Client streaming: requests in which the client streams data to the server. The
+  retryability of these requests is highly dependent on the type of data being
+  streamed. This document does not provide guidance for these requests.
+- Bidirectional: requests in which both the client streams to the server, and
+  the server streams to the client. The retryability of these requests is
+  highly dependent on the type of data being streamed. This document does not
+  provide guidance for these requests.
+
+### Retryable codes
+
+These codes **should** be retried for retryable requests:
+
+- `INTERNAL`: This generally means that some internal part of the system has
+failed. These are retryable under the expectation that the system will recover
+from failures.
+- `UNAVAILABLE`: These codes generally result from network hiccups, and are
+generally transient. They are retryable under the expectation that the
+connection will become available (soon).
+
+### Non-retryable codes
+
+These codes **should** never be retried for any request:
+
+- `OK`: The request succeeded.
```

Comment:
This might be better as a markdown table (if that's supported?)

Reply:
Markdown tables are supported. I am okay with this as-is, and am also okay with a table.

Reply:
Done https://github.com/jadekler/aip/blob/194/aip/0194.md